core/dbt/contracts/graph/nodes.py

```diff
@@ -546,19 +546,38 @@ def same_contract(self, old) -> bool:
             # A change, but not a breaking change
             return False
 
-        breaking_change_reasons = []
+        contract_enforced_disabled: bool = False
+        columns_removed: List[str] = []
+        column_type_changes: List[Tuple[str, str, str]] = []
         if old.contract.enforced is True and self.contract.enforced is False:
             # Breaking change: throw an error
             # Note: we don't have contract.checksum for current node, so build
             self.build_contract_checksum()
-            breaking_change_reasons.append("contract has been disabled")
+            contract_enforced_disabled = True
 
         if self.contract.checksum != old.contract.checksum:
-            # Breaking change, throw error
-            breaking_change_reasons.append("column definitions have changed")
-
-        if breaking_change_reasons:
-            raise (ModelContractError(reasons=" and ".join(breaking_change_reasons), node=self))
+            for key, value in sorted(old.columns.items()):
+                # Has this column been removed?
+                if key not in self.columns.keys():
+                    columns_removed.append(value.name)
+                # Has this column's data type changed?
+                elif value.data_type != self.columns[key].data_type:
+                    column_type_changes.append(
+                        (str(value.name), str(value.data_type), str(self.columns[key].data_type))
+                    )
+                # Otherwise, this was an additive change -- not breaking
+                else:
+                    continue
+
+        if contract_enforced_disabled or columns_removed or column_type_changes:
+            raise (
+                ContractBreakingChangeError(
+                    contract_enforced_disabled=contract_enforced_disabled,
+                    columns_removed=columns_removed,
+                    column_type_changes=column_type_changes,
+                    node=self,
+                )
+            )
         else:
             # no breaking changes
             return True
```

Comment:
It might just be a confusing comment - but this method should return True when there are _no_ changes (breaking or otherwise) to the contract, and False if there is a change to the contract but it is non-breaking. Are we actually handling the latter case?

Reply:
It does look like the logic isn't right here.  I think we need to keep track of adds and return False, so we need "elif columns_added, return False, else True.

Reply:
I'm taking a look at this logic now. It's genuinely tricky. I'll add more inline comments.

Reply:
I inverted the logic a bit, so now it's:
- Was the contract being enforced previously?
    - no &rarr; return `False` or `True`, but never `ContractBreakingChangeError`
- If so, do the checksums match up?
    - yes &rarr; `return True`
- If the checksums don't match up, are there any changes we consider breaking?
    - breaking &rarr; throw a `ContractBreakingChangeError` with details
    - non-breaking, still a change &rarr; `return False`
    
Let me know if you think that makes more or less sense than before